internal/ingest/ingest.go

```diff
@@ -678,30 +679,63 @@ func (ing *Ingester) ingestWorker() {
 	// It's assumed that the runIngestStep puts a mutex in this map.
-	ing.providersBeingProcessed[provider].Lock()
-	defer ing.providersBeingProcessed[provider].Unlock()
-
-	v, err := ing.ds.Get(context.Background(), datastore.NewKey(adProcessedPrefix+adCid.String()))
-	if err == nil && v[0] == byte(1) {
-		// We've process this ad already, so we know we've processed all earlier ads too.
-		return true
-	}
-	err = ing.ingestAd(publisher, adCid)
-	if err != nil {
-		log.Errorw("Error while ingesting ad.", "adCid", adCid, "publisher", publisher, "err", err)
+	ing.providersBeingProcessed[msg.provider].Lock()
+	defer ing.providersBeingProcessed[msg.provider].Unlock()
+
+	// Filter out ads that we've already processed and any earlier ads
+	splitAtIndex := len(msg.cids)
+	for i, adCid := range msg.cids {
+		// iterate latest to earliest
+		v, err := ing.ds.Get(context.Background(), datastore.NewKey(adProcessedPrefix+adCid.String()))
+		if err == nil && v[0] == byte(1) {
+			// We've process this ad already, so we know we've processed all
+			// earlier ads too. Break here and split at this index later. The cids
+			// before this index are newer and we haven't processed them, the cids
+			// after are older and we already have processed them.
+			splitAtIndex = i
+			break
+		}
 	}
+	yetToBeProcessed := msg.cids[:splitAtIndex]
+
+	log.Infow("Running worker on ad stack", "headAdCid", msg.cids[0], "publisher", msg.publisher, "numAdsToProcess", len(yetToBeProcessed))
+	for i := len(yetToBeProcessed) - 1; i >= 0; i-- {
+		// Note that we are iterating backwards here. Earliest to newest.
+		adCid := yetToBeProcessed[i]
+		err := ing.ingestAd(msg.publisher, adCid)
+
+		var adIngestErr adIngestError
+		if errors.As(err, &adIngestErr) {
+			switch adIngestErr.state {
+			case adIngestDecodingErr, adIngestMalformedErr, adIngestEntryChunkErr:
+				// These error cases are permament. e.g. if we try again later we will hit the same error. So we log and drop this error.
+				log.Errorw("Skipping ad because of a permanant error", "adCid", adCid, "err", err, "errKind", adIngestErr.state)
+				err = nil
+			}
+		}
 
-	return false
+		if err != nil {
+			log.Errorw("Error while ingesting ad. Bailing early, not ingesting later ads.", "adCid", adCid, "publisher", msg.provider, "err", err, "adsLeftToProcess", i+1)
+
+			// Tell anyone who's waiting that the sync finished for this head because we errored.
+			// TODO(mm) would be better to propagate the error.
+			ing.inEvents <- legs.SyncFinished{Cid: yetToBeProcessed[0], PeerID: msg.publisher}
+			return
+		}
+
+		if markErr := ing.markAdProcessed(msg.publisher, adCid); markErr != nil {
+			log.Errorw("Failed to mark ad as processed", "err", markErr)
+		} else {
+			stats.Record(context.Background(), metrics.AdSyncedCount.M(1))
+		}
+		// Distribute the legs.SyncFinished notices to waiting Sync calls.
+		ing.inEvents <- legs.SyncFinished{Cid: adCid, PeerID: msg.publisher}
+	}
```

Comment:
```suggestion
		// Cleanup add from datastore.
		err = ing.ds.Delete(context.Background(), dsKey(adCid.String()))
		if err != nil {
			log.Errorw("Failed to clean up ad from datastore", "err", err)
		}
	}
```
I think this can be cleaned up now that it is completely ingested.  Maybe it is not even necessary to store the ad in the datastore at all, and use only the adCache instead.